text/0102-offchain-parachain-runtime-upgrades.md

```diff
@@ -0,0 +1,354 @@
+
+1. Fetching can happen over a longer period of time with low priority. E.g. if
+   we waited for the PVF at the very first avaialbility distribution, this might
+   actually affect liveness of other chains on the same core. Distributing
+   megabytes of data to a thousand validators, might take a bit. Thus this helps
+   isolating parachains from each other.
+2. By configuring the initial counter value we can affect how much an upgrade
+   costs. E.g. forcing the parachain to produce 10 blocks, means 10x the cost
+   for issuing an update. If too frequent upgrades ever become a problem for the
+   system, we have a knob to make them more costly.
+
+### On-chain code upgrade process
+
+First when a candidate is backed we need to make the new hash available
+(together with a counter) via a
+runtime API so validators in availability distribution can check for it and
+fetch it if changed (see previous section). For performance reasons, I think we
+should not do an additional call, but replace the [existing one](https://github.com/paritytech/polkadot-sdk/blob/d2fd53645654d3b8e12cbf735b67b93078d70113/polkadot/node/subsystem-util/src/runtime/mod.rs#L355) with one containing the new additional information (Option<(Hash, Counter)>).
+
+Once the candidate gets included (counter 0), the hash is given to pre-checking
+and only after pre-checking succeeded (and a full session passed) it is finally
+enacted and the parachain can switch to the new code. (Same process as it used
+to be.)
+
+### Handling new validators
+#### Backers
+
+If a backer receives a collation for a parachain it does not yet have the code
+as enacted on chain (see "On-chain code upgrade process"), it will use above
+request/response protocol to fetch it from whom it received the collation.
+
+#### Availablity Distribution
+
+Validators in availability distribution will be changed to only sign a `1` in
+the bitfield of a candidate if they not only have the chunk, but also the
+currently active PVF. They will fetch it from backers in case they don't have it
+yet.
+
+### How do other parties get hold of the PVF?
+
+Two ways:
+
+1. Discover collators via [relay chain DHT](https://github.com/polkadot-fellows/RFCs/pull/8) and request from them: Preferred way,
+   as it is less load on validators.
+2. Request from validators, which will serve on a best effort basis.
+
+### Pruning
+
+We covered how validators get hold of new code, but when can they prune old ones?
+In principle it is not an issue, if some validors prune code, because:
+
+1. We changed it so that a candidate is not deemed available if validators were
+   not able to fetch the PVF.
+2. Backers can always fetch the PVF from collators as part of the collation
+   fetching.
+
+But the majority of validators should always keep the latest code of any
+parachain and only prune the previous one, once the first candidate using the
+new code got finalized. This ensures that disputes will always be able to
+resolve.
```

Comment:
Yeah 1 is an improvement here, previously I'd envisions parachains doing code reuploads once per day, just so the code stays in availability